### PR TITLE
feat(android): use device serial for `deviceName`

### LINKED ISSUE #4042: Use hardware serial for Android deviceName
There will be lots of duplicate models out there, so it would be a good idea to use a device serial (or some other differentiator) + device name as the device name on Android.

The problem this solves is making it easier to tell clients apart at-a-glance in the portal.

### PR DESCRIPTION
**Pull request by jasonboukheir**
Fixes #4042 

The serial number of the device is blocked behind a permission. There's a couple ways we can go about this:

-----

### (1) Ask the user to (optionally) grant the permission

When we show the grant VPN permission activity, we also mention the optional READ_PRIVILEGED_PHONE_STATE permission. Here, the user can decide to grant it or not, and if they decide not to, they can grant it in the future in the app settings. When the permission is not granted, the `deviceName` falls back to the `Build.MODEL`

### (2) Force the user to grant the permission

We keep asking them to grant the permission in the splash view. `deviceName` is always the serial number of the device.

### (3) Let MDM grant the permission

We don't provide a UI to grant the permission in the application. Instead, the `deviceName` is the `Build.MODEL` by default, unless advanced users or admins using MDM set the permission, in which case it's the serial number of the device.

### (4) Let MDM set a custom/override device name

This could be an alternative to (3) if it is easier for customers using MDM software to manage it this way. Though I doubt it...

-----

Going with option (3) is safe, and the other options can be added incrementally in the future. However, it requires communicating to the customer that they should set this permission for the `deviceName` to be the serial of the device. That's not a problem yet, since the relevant customer is using MDM to manage the app; it's trivial to set this permission via that UI.

If we did want to show this permission to the user, I think option (1) is most likely going to be better than option (2) in most cases.

### CODE REVIEW

**Comment by jamilbk** [comment]
@jasonboukheir Thanks for summarizing our options, that makes things much more clear! I'll link the customer to this PR for input on which method is best. To avoid privacy and data concerns that affect all customers, I'm thinking 3 or 4.

**Comment by AndrewDryga** [comment]
I like the option 3. And I think we should not reuse Device Name for it but better we add a separate optional field for it that will be displayed in the portal.

**Comment by jamilbk** [comment]
> I like the option 3. And I think we should not reuse Device Name for it but better we add a separate optional field for it that will be displayed in the portal.

Option 4 would be more consistent with how we allow users to set their own `FIREZONE_NAME` which we allow for the gateway and headless client. If the MDM solution supports setting that en-masse then that would work best since it wouldn't require any portal changes and give them control over what they see in the portal.

The underlying use case here is being able to quickly filter thousands of Clients in the table view by some more unique identifier than the device's model number. It could be a serial in this case, but for other Clients it could be a MAC address, Firebase.installationId, etc.

Keeping `FIREZONE_NAME` agnostic to what it represents may be more flexible and prevent us from having to add a specific metadata field for each piece of information we may or may not be able to retrieve from a certain platform.

**Comment by jamilbk** [comment]
~~Option 4 would also allow us from having to declare this information as "collected" in our privacy statements for the Google Play Store, since it's provided by the customer.~~

Edit: On second thought, we're still saving it, so I'll need to think about the privacy implications.

**Comment by jamilbk** [comment]
@jasonboukheir Customer input is that they can set the field to a variable, so they can override to a value of their choosing. Let's go with Option (4) which prevents us from needing to request additional permissions.

Keep in mind if the MDM config field is not set it will be `null` and not the empty string, so you need to check both of those before deciding to fallback to the build.MODEL

**Comment by tobiasmcnulty** [comment]
Option 4 is a great fit for our use case; TinyMDM at least supports variables in the managed configuration interface, so we can populate the name with the serial number or whatever other variables me might want to in the future.

If this isn't too complicated, Option 3 would work well as a default value, i.e., if no name is provided via the MDM configuration, you could call back to the serial number or at worst the model. TinyMDM indeed gives admins the option to pre-approve permissions that would otherwise be requested by the application.

Thanks so much for the detailed write-up and options!

**Comment by jasonboukheir** [comment]
Thank you for the feedback, everyone! Combo of (3) and (4) it is :)

**Comment by jasonboukheir** [comment]
Screenshot 2024-03-20 at 10 15 30 PM

Testing it out with MDM. Looks like it works!

There is another issue that is related to this #4181. We should make sure the tunnel is restarted when this mdm property changes as well. Merge this first, then I'll make sure to add the other property in that PR.

File: kotlin/android/app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -62,4 +62,9 @@
 		If this list is empty, no applications are disallowed. Either "Allowed Applications" OR
 		"Disallowed Applications" may be set *but not* both.
 	</string>
+	<string name="config_device_name_title">Device Name</string>
+	<string name="config_device_name_description">
+		The name of the device. This is used to identify the device in the Firezone console.
```
Comment (jamilbk) [inline kotlin/android/app/src/main/res/values/strings.xml]:
```suggestion
		The name of the device. This is used to identify the device in the admin portal.
```

File: kotlin/android/app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -62,4 +62,9 @@
...
+		If unset, then the device name will be the device\'s model name.
```
Comment (jamilbk) [inline kotlin/android/app/src/main/res/values/strings.xml]:
```suggestion
		If unset, device\'s model name will be used.
```

**Review by jamilbk — APPROVED** [review]

**Pull request merged**